File: interface/types/events.ts
// V7 Event Spine — TypeScript types for mission runs and events.
// Mirrors core/pkg/protocol/events.go and core/internal/runs/manager.go.
// Used by RunTimeline.tsx, EventCard.tsx, ViewChain.tsx (Team E components).

export type EventType =
  | 'mission.started'
  | 'mission.completed'
  | 'mission.failed'
  | 'mission.cancelled'
  | 'team.spawned'
  | 'team.stopped'
  | 'agent.started'
  | 'agent.stopped'
  | 'tool.invoked'
  | 'tool.completed'
  | 'tool.failed'
  | 'artifact.created'
  | 'memory.stored'
  | 'memory.recalled'
  | 'trigger.fired'
  | 'trigger.skipped'
  | 'scheduler.tick';

export type EventSeverity = 'debug' | 'info' | 'warn' | 'error';

export type RunStatus = 'pending' | 'running' | 'completed' | 'failed';

// MissionEventEnvelope is the authoritative audit record for a mission execution event.
// Returned by GET /api/v1/runs/{id}/events.
export interface MissionEventEnvelope {
  id: string;
  run_id: string;
  tenant_id: string;
  event_type: EventType;
  severity: EventSeverity;
  source_agent?: string;
  source_team?: string;
  payload?: Record<string, unknown>;
  audit_event_id?: string;
  emitted_at: string; // ISO 8601
}

// MissionRun is a single execution instance of a mission definition.
// Returned by GET /api/v1/runs/{id}/chain (chain array).
export interface MissionRun {
  id: string;
  mission_id: string;
  tenant_id: string;
  status: RunStatus;
  run_depth: number;
  parent_run_id?: string;
  started_at: string; // ISO 8601
  completed_at?: string; // ISO 8601
  metadata?: Record<string, unknown>;
}

// RunChainResponse is returned by GET /api/v1/runs/{id}/chain.
export interface RunChainResponse {
  run_id: string;
  mission_id: string;
  chain: MissionRun[];
}

// Color mapping for event types — used by RunTimeline and EventCard.
export const EVENT_TYPE_COLORS: Record<string, string> = {
  'mission.started':   '#10b981', // cortex-success green
  'mission.completed': '#10b981',
  'mission.failed':    '#ef4444', // red
  'mission.cancelled': '#f59e0b', // amber
  'team.spawned':      '#06b6d4', // cortex-primary cyan
  'team.stopped':      '#71717a', // muted
  'agent.started':     '#06b6d4',
  'agent.stopped':     '#71717a',
  'tool.invoked':      '#8b5cf6', // violet
  'tool.completed':    '#10b981',
  'tool.failed':       '#ef4444',
  'artifact.created':  '#f59e0b',
  'memory.stored':     '#06b6d4',
  'memory.recalled':   '#06b6d4',
  'trigger.fired':     '#f59e0b',
  'trigger.skipped':   '#71717a',
  'scheduler.tick':    '#71717a',
};

export const SEVERITY_COLORS: Record<EventSeverity, string> = {
  debug: '#71717a',
  info:  '#06b6d4',
  warn:  '#f59e0b',
  error: '#ef4444',
};

File: core/pkg/protocol/events.go
package protocol

import (
	"context"
	"time"
)

// EventType classifies a mission event in the persistent audit trail.
// All 17 event types are defined here as constants to prevent typos.
type EventType string

const (
	// Mission lifecycle
	EventMissionStarted   EventType = "mission.started"
	EventMissionCompleted EventType = "mission.completed"
	EventMissionFailed    EventType = "mission.failed"
	EventMissionCancelled EventType = "mission.cancelled"

	// Team lifecycle
	EventTeamSpawned EventType = "team.spawned"
	EventTeamStopped EventType = "team.stopped"

	// Agent lifecycle
	EventAgentStarted EventType = "agent.started"
	EventAgentStopped EventType = "agent.stopped"

	// Tool execution (ReAct loop)
	EventToolInvoked   EventType = "tool.invoked"
	EventToolCompleted EventType = "tool.completed"
	EventToolFailed    EventType = "tool.failed"

	// Artifact lifecycle
	EventArtifactCreated EventType = "artifact.created"

	// Memory (MCP memory server)
	EventMemoryStored   EventType = "memory.stored"
	EventMemoryRecalled EventType = "memory.recalled"

	// Automation (Team B/C — registered here for completeness)
	EventTriggerFired   EventType = "trigger.fired"
	EventTriggerSkipped EventType = "trigger.skipped"
	EventSchedulerTick  EventType = "scheduler.tick"
)

// EventSeverity classifies the operational significance of an event.
type EventSeverity string

const (
	SeverityDebug EventSeverity = "debug"
	SeverityInfo  EventSeverity = "info"
	SeverityWarn  EventSeverity = "warn"
	SeverityError EventSeverity = "error"
)

// MissionEventEnvelope is the authoritative audit record for a mission execution event.
// Every action in a mission run MUST produce one of these records (V7 Event Spine rule).
// This is the PERSISTENT record. The CTS envelope is the real-time transport signal.
// CTS payload includes mission_event_id to link signals back to this record.
type MissionEventEnvelope struct {
	ID           string                 `json:"id"`
	RunID        string                 `json:"run_id"`
	TenantID     string                 `json:"tenant_id"`
	EventType    EventType              `json:"event_type"`
	Severity     EventSeverity          `json:"severity"`
	SourceAgent  string                 `json:"source_agent,omitempty"`
	SourceTeam   string                 `json:"source_team,omitempty"`
	Payload      map[string]interface{} `json:"payload,omitempty"`
	AuditEventID string                 `json:"audit_event_id,omitempty"`
	EmittedAt    time.Time              `json:"emitted_at"`
}

// EventEmitter allows optional event emission without creating import cycles.
// The interface lives in protocol because swarm and server both import protocol.
// Implemented by internal/events.Store; nil receiver = silent mode (no panic).
type EventEmitter interface {
	Emit(ctx context.Context, runID string, eventType EventType, severity EventSeverity,
		sourceAgent, sourceTeam string, payload map[string]interface{}) (string, error)
}

// RunsManager allows optional run tracking without creating import cycles.
// The interface lives in protocol because swarm imports protocol.
// Implemented by internal/runs.Manager; nil receiver = no run tracking.
type RunsManager interface {
	CreateRun(ctx context.Context, missionID string) (string, error)
	UpdateRunStatus(ctx context.Context, runID string, status string) error
}

// ConversationTurnData holds all fields for a single conversation turn.
// Used by ConversationLogger.LogTurn to persist full-fidelity agent transcripts.
type ConversationTurnData struct {
	RunID          string                 `json:"run_id,omitempty"`     // nullable: standing-team chats
	SessionID      string                 `json:"session_id"`
	TenantID       string                 `json:"tenant_id"`
	AgentID        string                 `json:"agent_id"`
	TeamID         string                 `json:"team_id,omitempty"`
	TurnIndex      int                    `json:"turn_index"`
	Role           string                 `json:"role"`                 // system|user|assistant|tool_call|tool_result|interjection
	Content        string                 `json:"content"`
	ProviderID     string                 `json:"provider_id,omitempty"`
	ModelUsed      string                 `json:"model_used,omitempty"`
	ToolName       string                 `json:"tool_name,omitempty"`
	ToolArgs       map[string]interface{} `json:"tool_args,omitempty"`
	ParentTurnID   string                 `json:"parent_turn_id,omitempty"` // links tool_result → tool_call
	ConsultationOf string                 `json:"consultation_of,omitempty"`
}

// ConversationLogger allows optional conversation turn persistence.
// Analogous to EventEmitter — agents receive it via SetConversationLogger().
// nil = silent mode (no panic, no recording).
// Implemented by internal/conversations.Store.
type ConversationLogger interface {
	LogTurn(ctx context.Context, turn ConversationTurnData) (string, error)
}
